### PR TITLE
Telemetry: Lvm exporter for Prometheus

### PR DESCRIPTION
**Pull request by NajmudheenCT**
**What this PR does / why we need it**:
This PR is the first version of lvm exporter for Prometheus .
lvm exporter creates and http end point for Prometheus/third party to scrape/consume metrics.This uses openSDS lvm metric driver code to collect metrics and expose an endpoint .
This is an independent binary where user has to provide the port number for endpoint as command line parameter.
This port number can be configured in Prometheus config to scrape from lvm_exporter .
**Which issue this PR fixes** 
NA
**Special notes for your reviewer**:
dep ensure has been run to update prometheus and other dependency packages.
This operation downloaded so many files to /vendor folder
**Release note**:

### CODE REVIEW

**Comment by NajmudheenCT** [comment]
@skdwriting @xing-yang @wisererik @leonwanghui @rhsakarpos

File: contrib/drivers/lvm/lvm_metrics.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+	for _, element := range metricsList {
+		val, _ := strconv.ParseFloat(metricMap[element], 64)
+		//Todo: See if association  is required here, resource discovery could fill this information
+		associatorMap := make(map[string]string)
+
+		metricValue := model.Metric{
+			Timestamp: getCurrentUnixTimestamp(),
+			Value:     val,
+		}
+		metricValues := make([]model.Metric, 0)
```
Comment (leonwanghui) [inline contrib/drivers/lvm/lvm_metrics.go]: Don't use `[]model.Metric`, rather change it to `[]*model.Metric`.

File: contrib/drivers/lvm/metrics_cli.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+import (
+	log "github.com/golang/glog"
+	"github.com/opensds/opensds/pkg/utils/exec"
+	"regexp"
+	"strings"
+)
+
+const (
+	sarNotEnabledOut = "Please check if data collecting is enabled"
```
Comment (leonwanghui) [inline contrib/drivers/lvm/metrics_cli.go]: Please change the first letter to lower case for error msg.

File: contrib/drivers/lvm/metrics_cli.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		BaseExecuter: exec.NewBaseExecuter(),
+		RootExecuter: exec.NewRootExecuter(),
+	}, nil
+}
+
+func (c *MetricCli) execute(cmd ...string) (string, error) {
+	return c.RootExecuter.Run(cmd[0], cmd[1:]...)
+}
+
+func is_sar_enabled(out string) bool {
```
Comment (leonwanghui) [inline contrib/drivers/lvm/metrics_cli.go]: s/is_sar_enabled/isSarEnabled

Comment (leonwanghui) [reply contrib/drivers/lvm/metrics_cli.go]: What do you mean by using `out` here?

File: contrib/drivers/lvm/metrics_cli.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+	if strings.Contains(string(out), sarNotEnabledOut) || strings.Contains(string(out), cmdNotFound) {
+
+		return false
+	}
+
+	return true
+}
+
+//	Function to parse sar and iostat command output
```
Comment (leonwanghui) [inline contrib/drivers/lvm/metrics_cli.go]: Please update the style of comments.

File: contrib/drivers/lvm/metrics_cli.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+				returnMap[metric] = val
+				returnMap["InstanceName"] = columns[metricMap["InstanceID"]]
+
+			}
+		}
+
+	}
+}
+
+//	CollectMetrics function is call the cli for metrics collection. This will be invoked  by lvm metric driver
```
Comment (leonwanghui) [inline contrib/drivers/lvm/metrics_cli.go]: Same here

File: pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+			"Shows Utilization in percentage",
+			volumeLabel, nil,
+		),
+	}
+
+}
+
+//Describe function.
+//It essentially writes all descriptors to the prometheus desc channel.
+func (collector *lvmCollector) Describe(ch chan<- *prometheus.Desc) {
```
Comment (leonwanghui) [inline pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go]: I think `c` is ok compared to `collector`.

File: pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+func validateCliArg(arg1 string) string {
+	num, err := strconv.Atoi(arg1)
+	if (err != nil) || (num > 65535) {
+
+		fmt.Println("Please enter a valid port number")
+		os.Exit(1)
+	}
+	return arg1
+}
+func main() {
```
Comment (leonwanghui) [inline pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go]: I'm afraid I don't understand why `main` method is defined here.

**Comment by NajmudheenCT** [comment]
> Please don't import any binary files in the source code.

deleted binary from repository. Will add how to generate binary steps in documents.

File: contrib/drivers/lvm/metrics_cli.go (hunk 3 above)
Comment (NajmudheenCT) [reply contrib/drivers/lvm/metrics_cli.go]: > s/is_sar_enabled/isSarEnabled

Changed to camel case

Comment (NajmudheenCT) [reply contrib/drivers/lvm/metrics_cli.go]: > What do you mean by using `out` here?

out is used here to check whether the output of command contains ""Please check if data collecting is enabled" , this means sar is not enabled

File: pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go (hunk 7 above)
Comment (NajmudheenCT) [reply pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go]: lvm_exporter is an independent process which user can start if required.

**Comment by NajmudheenCT** [comment]
> I suggest you choose a specified version of added vendor packages, please change it in `Gopkg.toml`.

@leonwanghui  updated .toml file,
Thanks

File: contrib/drivers/lvm/lvm_metrics.go [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+import (
+	log "github.com/golang/glog"
+	"github.com/opensds/opensds/pkg/model"
+	"gopkg.in/yaml.v2"
+	"strconv"
+	"time"
+)
+
+/*const (
```
Comment (wisererik) [inline contrib/drivers/lvm/lvm_metrics.go]: remove unused code

File: contrib/drivers/lvm/lvm_metrics.go [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+func getMetricToUnitMap() map[string]string {
+
+	//construct metrics to value map
+	var configs Configs
+	//Read supported metric list from yaml config
+	//source, err := ioutil.ReadFile(ConfFile)
+	//Todo: Move this to read from file
+	source := []byte(data)
+	/*if err != nil {
+	log.Errorf("Not able to read file  %s  %v", ConfFile, err)
```
Comment (wisererik) [inline contrib/drivers/lvm/lvm_metrics.go]: remove it

File: contrib/drivers/lvm/lvm_metrics.go [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+//	supportedMetrics -> list of supported metrics
+func (d *MetricDriver) ValidateMetricsSupportList(metricList []string, resourceType string) (supportedMetrics []string, err error) {
+	var configs Configs
+
+	//Read supported metric list from yaml config
+	//source, err := ioutil.ReadFile(ConfFile)
+	//Todo: Move this to read from file
+	source := []byte(data)
+	/*if err != nil {
+		log.Errorf("Not able to read file  %s%v", ConfFile, err)
```
Comment (wisererik) [inline contrib/drivers/lvm/lvm_metrics.go]: remove it

File: contrib/drivers/lvm/lvm_metrics.go (hunk 8 above)
Comment (NajmudheenCT) [reply contrib/drivers/lvm/lvm_metrics.go]: @wisererik , removed , Thanks

**Pull request merged**